pkg/clusteragent/clusterchecks/stores.go

```diff
@@ -102,3 +116,71 @@ func (s *nodeStore) removeConfig(digest string) {
 	delete(s.digestToConfig, digest)
 	dispatchedConfigs.WithLabelValues(s.name).Dec()
 }
+
+// AddRunnerStats stores runner stats for a check
+// The nodeStore handles thread safety for this public method
+func (s *nodeStore) AddRunnerStats(checkID string, stats types.CLCRunnerStats) {
+	s.Lock()
+	defer s.Unlock()
+	s.clcRunnerStats[checkID] = stats
+}
+
+// RemoveRunnerStats deletes runner stats for a check
+// The nodeStore handles thread safety for this public method
+func (s *nodeStore) RemoveRunnerStats(checkID string) {
+	s.Lock()
+	defer s.Unlock()
+	_, found := s.clcRunnerStats[checkID]
+	if !found {
+		log.Debugf("unknown check ID %s, skipping", checkID)
+		return
+	}
+	delete(s.clcRunnerStats, checkID)
+}
+
+// GetRunnerStats returns the runner stats of a given check
+// The nodeStore handles thread safety for this public method
+func (s *nodeStore) GetRunnerStats(checkID string) types.CLCRunnerStats {
+	s.RLock()
+	defer s.RUnlock()
+	stats, found := s.clcRunnerStats[checkID]
+	if !found {
+		log.Debugf("unknown check ID %s", checkID)
+	}
+	return stats
+}
+
+// GetBusyness calculates busyness of the node
+// The nodeStore handles thread safety for this public method
+func (s *nodeStore) GetBusyness(busynessFunc func(avgExecTime, mSamples int) int) int {
+	s.RLock()
+	defer s.RUnlock()
+	busyness := 0
+	for _, stats := range s.clcRunnerStats {
+		busyness += busynessFunc(stats.AverageExecutionTime, stats.MetricSamples)
+	}
+	return busyness
+}
+
+// GetMostWeightedCheck returns the check with the most weight on the node
+// The nodeStore handles thread safety for this public method
+func (s *nodeStore) GetMostWeightedCheck(busynessFunc func(avgExecTime, mSamples int) int) (string, int, error) {
+	s.RLock()
+	defer s.RUnlock()
+	if len(s.clcRunnerStats) == 0 {
+		log.Debugf("Node %s has no check stats", s.name)
+		return "", -1, fmt.Errorf("node %s has no check stats", s.name)
+	}
+	firstItr := true
+	checkID := ""
```

Comment:
can we have an issue if the function return `checkID=""` ?

Reply:
hypothetically there is two cases where this function "can" return an empty `checkID`
- if it doesn't access the loop `for id, stats := range s.clcRunnerStats`, this is impossible because the function returns an error if len(s.clcRunnerStats) == 0 (the error is correctly handled by the rebalancing logic)
- if we have an empty key in `s.clcRunnerStats` which less likely to happen (it depends on runner stats server), but to make it safer, I'll make sure the CLC runner client used by the DCA ignore empty check IDs if any, and add a test case for it

Does it make sense ?

Reply:
ok